Add Android build to CI

junhochoi opened the pull request:
Add android target, build 4 platforms. Since android
language support doesn't include rust and cmake version is too
old, install it separately.

ghedo commented on .travis.yml:
```diff
@@ -69,6 +71,36 @@ matrix:
+      - unzip -q ndk.zip -d $HOME
+      - export ANDROID_NDK_HOME=$HOME/android-ndk-$NDK_VER
+      # Install latest cmake. preinstalled version is too old (3.2.2)
+      - CMAKE_URL=https://cmake.org/files/v3.15/cmake-3.15.4-Linux-x86_64.tar.gz
```
What version of cmake is required? The ubuntu bionic image has 3.12 https://docs.travis-ci.com/user/reference/bionic/#compilers-and-build-toolchain if we could just use that image we wouldn't need to install cmake manually.

junhochoi replied:
We need cmake 3.6.0 and later otherwise boringssl build will fail. For Android build we need to use `trusty` image. https://docs.travis-ci.com/user/languages/android/
And `trusty` image has cmake 3.2.2 (`/usr/bin/cmake`) so had to install newer version. This is outside of NDK build, so NDK cmake doesn't help.

junhochoi replied:
Ok. I was able to install `cmake` using `sdkmanager`, replacing this installation procedure.

ghedo commented on .travis.yml:
```diff
@@ -69,6 +71,36 @@ matrix:
+      - export PATH=$HOME/.cargo/bin:$PATH
+      - rustup default stable
+      - rustup target add aarch64-linux-android arm-linux-androideabi armv7-linux-androideabi i686-linux-android
+      # Install android ndk
```
Can we use the `sdkmanager` command for this? https://github.com/android/ndk-samples/blob/master/.travis.yml#L18

junhochoi replied:
Here I am reusing how other rust project building with ndk (e.g. libc, ring).  I will take a look again.

junhochoi replied:
Now we use `sdkmanager` for `cmake`. I tested installing `ndk-bundle` with `sdkmanager` but it fails... I think it's because `ndk-bundle` coming with this `sdkmanager` is NDK r20 and we are not still compatible with this. There is toolchain script change somewhere between 13b and 20, breaking backward compatibility with old NDK. r13b was chosen based on current practice. If we want to be compatible with NDK 20 then we need a separate work.

ghedo commented on .travis.yml:
```diff
@@ -69,6 +71,36 @@ matrix:
+     install:
+      # Install rust manually
+      - curl https://build.travis-ci.org/files/rustup-init.sh -sSf | sh -s -- -y --default-toolchain stable
+      - export PATH=$HOME/.cargo/bin:$PATH
```
Should probably just put this in an `env` section https://docs.travis-ci.com/user/environment-variables/

junhochoi replied:
We still need to install `rust` this way unless we have a way to have an environment with 2 language (android and rust).

ghedo commented on .travis.yml:
```diff
@@ -69,6 +71,36 @@ matrix:
+      - rustup default stable
+      - rustup target add aarch64-linux-android arm-linux-androideabi armv7-linux-androideabi i686-linux-android
+      # Install android ndk
+      - NDK_VER=r13b
```
Same about `env` https://docs.travis-ci.com/user/environment-variables/

junhochoi replied:
Moved some constants for NDK and its url to env/global section.

ghedo commented:
Reabased on master, and also moved the `env` declaration to the actual build that uses those.

junhochoi commented:
> Reabased on master, and also moved the `env` declaration to the actual build that uses those.

Thanks. I think it's good to go.

junhochoi commented:
> > Reabased on master, and also moved the `env` declaration to the actual build that uses those.
> 
> Thanks. I think it's good to go.

Moved NDK_URL inside build script. It's too long so not looking good at CI matrix table.
